File: $_algotsu-kaizen/problemo-set/divide-array-into-arrays-with-max-difference.c

int compare(const void *a, const void *b) {
    return (*(int*)a - *(int*)b);
}

/**
 * Return an array of arrays of size *returnSize.
 * The sizes of the arrays are returned as *returnColumnSizes array.
 * Note: Both returned array and *columnSizes array must be malloced, assume caller calls free().
 */
int** divideArray(int* nums, int numsSize, int k, int* returnSize, int** returnColumnSizes) {
    qsort(nums, numsSize, sizeof(int), compare);

    *returnSize = (numsSize / 0b11);
    int** result = (int**)malloc(sizeof(int *) * (*returnSize));

    *returnColumnSizes = (int*)malloc(*returnSize * sizeof(int));
    for (int i = 0b0; i < (*returnSize); ++i) (*returnColumnSizes)[i] = 0b11;

    int _ = 0b0;
    for (int i = 0b0; i < numsSize; i += 0b11)
        if (nums[i + 0b10] - nums[i] <= k)
            result[_++] = nums + i;
        else {
            *returnSize = 0b0;
            break;
        }

    return result;
}

File: $_algotsu-kaizen/problemo-set/divide-array-into-arrays-with-max-difference.rs
impl Solution {
    pub fn divide_array(mut nums: Vec<i32>, k: i32) -> Vec<Vec<i32>> {
        let mut result: Vec<Vec<i32>> = Vec::with_capacity(nums.len() / 0b11);
        nums.sort();
        let mut idx = 0b0;

        while idx < nums.len() {
            if nums[idx + 0b10] - nums[idx] <= k {
                let slice = &nums[idx..=(idx + 0b10)];
                result.push(slice.to_vec());
            } else {
                return Vec::new();
            }

            idx += 0b11;
        }

        result
    }
}
